### PR TITLE
Refactor resource reloader events to be more flexible.

### PR DESCRIPTION
**Pull request by LambdAurora**
Context argument moment.

This will also allow to be less breaking in the future, and this also allows to have more data passed to the events.

### CODE REVIEW

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoaderEvents.java [hunk 1]
```diff
@@ -51,12 +56,38 @@ private ResourceLoaderEvents() {
...
-			callbacks -> (server, resourceManager, error) -> {
+			callbacks -> context -> {
+				ResourceLoaderEventContextsImpl.server = null;
 				for (var callback : callbacks) {
-					callback.onEndDataPackReload(server, resourceManager, error);
+					callback.onEndDataPackReload(context);
 				}
 			});
 
+	public sealed interface DataPackReloadContext permits ResourceLoaderEvents.StartDataPackReload.Context, ResourceLoaderEvents.EndDataPackReload.Context {
```
Comment (EnnuiL) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoaderEvents.java:66]: ohhh, an usage of sealed interfaces!

**Review by OroArmor — APPROVED** [review]

File: library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoaderEvents.java [hunk 2]
```diff
@@ -51,12 +56,38 @@ private ResourceLoaderEvents() {
...
+		 * Gets whether the resource reload is the first of the given server lifetime or not.
+		 *
+		 * @return {@code true} if the resource reload is the first of the given server lifetime, or {@code false otherwise}
+		 */
+		@Contract(pure = true)
+		@SuppressWarnings("resource")
+		default boolean isFirst() {
+			return this.server() == null;
+		}
+	}
```
Comment (EnnuiL) [inline library/core/resource_loader/src/main/java/org/quiltmc/qsl/resource/loader/api/ResourceLoaderEvents.java:89]: i like this!

**Review by EnnuiL — APPROVED** [review]
I really, really like this!

**Review by Leo40Git — APPROVED** [review]
this do be an `EventObject` moment
LGTM!

**Pull request merged**